test/test_linalg.py

```diff
@@ -1347,6 +1347,90 @@ def check(equation, operands, regex, exception=RuntimeError):
         check('a, ba', [x, y], r'operands do not broadcast with remapped shapes \[original->remapped\]: '
                                r'\[2\]->\[1, 2\] \[2, 3\]->\[2, 3\]')
 
+
+    @skipCUDAIfNoMagma
+    @skipCPUIfNoLapack
+    @dtypes(torch.float, torch.double, torch.cfloat, torch.cdouble)
+    def test_qr_reduced(self, device, dtype):
+        t = torch.randn((7, 5), device=device, dtype=dtype)
+        np_t = t.cpu().numpy()
+
+        exp_q, exp_r = np.linalg.qr(np_t)
+        q, r = torch.linalg.qr(t)
+        self.assertEqual(q, exp_q)
+        self.assertEqual(r, exp_r)
+
+        exp_q, exp_r = np.linalg.qr(np_t, mode='reduced')
+        q, r = torch.linalg.qr(t, mode='reduced')
+        self.assertEqual(q, exp_q)
+        self.assertEqual(r, exp_r)
+
+    @skipCUDAIfNoMagma
+    @skipCPUIfNoLapack
+    @dtypes(torch.float, torch.double, torch.cfloat, torch.cdouble)
+    def test_qr_complete(self, device, dtype):
+        t = torch.randn((7, 5), device=device, dtype=dtype)
+        np_t = t.cpu().numpy()
+
+        exp_q, exp_r = np.linalg.qr(np_t, mode='complete')
+        q, r = torch.linalg.qr(t, mode='complete')
+        self.assertEqual(q, exp_q)
+        self.assertEqual(r, exp_r)
+
+    @skipCUDAIfNoMagma
+    @skipCPUIfNoLapack
+    @dtypes(torch.float, torch.double, torch.cfloat, torch.cdouble)
+    def test_qr_r(self, device, dtype):
+        t = torch.randn((7, 5), device=device, dtype=dtype)
+        np_t = t.cpu().numpy()
+        # note: numpy returns only r, while we return a tuple q, r where q is
+        # an empty tensor
+        exp_r = np.linalg.qr(np_t, mode='r')
+        q, r = torch.linalg.qr(t, mode='r')
+        # check that q is empty
+        assert q.shape == (0,)
+        assert q.dtype == t.dtype
+        assert q.device == t.device
+        # check r
+        self.assertEqual(r, exp_r)
+
+    @skipCUDAIfNoMagma
+    @skipCPUIfNoLapack
+    @dtypes(torch.float)
+    def test_qr_empty(self, device, dtype):
+        def check(shape):
+            t = torch.empty((5, 0), dtype=dtype, device=device)
+            np_t = t.cpu().numpy()
+            exp_q, exp_r = np.linalg.qr(np_t)
+            q, r = torch.linalg.qr(t, mode='reduced')
+            self.assertEqual(q, exp_q)
+            self.assertEqual(r, exp_r)
+            q, r = torch.linalg.qr(t, mode='r')
+            assert q.shape == (0,)
+            self.assertEqual(r, exp_r)
+
+        check((5, 0))
+        check((0, 5))
+
+    @skipCUDAIfNoMagma
+    @skipCPUIfNoLapack
+    @dtypes(torch.float, torch.double, torch.cfloat, torch.cdouble)
+    def test_qr_out(self, device, dtype):
```

Comment:
Instead of testing out separately, I suggest trying to include it in the tests for each mode.